graphql_compiler/query_pagination/__init__.py

```diff
@@ -63,36 +75,32 @@ def paginate_query(schema_info, query_string, parameters, page_size):
 
     Args:
         schema_info: QueryPlanningSchemaInfo
-        schema_graph: SchemaGraph instance.
-        statistics: Statistics object.
-        query_string: str, valid GraphQL query to be paginated.
+        query: QueryStringWithParameters
         parameters: dict, parameters with which query will be estimated.
         page_size: int, describes the desired number of result rows per page.
 
     Returns:
         tuple containing queries for going over the original query in a paginated fashion:
             - QueryStringWithParameters namedtuple, query expected to return roughly a page of
               result data of the original query.
-            - QueryStringWithParameters namedtuple or None. If the given query was estimated to
-              return more than a page of results, this element is a QueryStringWithParameters
-              namedtuple describing a query for the rest of the result data. Otherwise, this element
-              is None.
+            - Tuple of QueryStringWithParameters, describing queries that return the rest of the
+              result data of the original query. If the original query is expected to return only
+              a page or less of results, then this element will be an empty tuple.
+            - Tuple of PaginationAdvisories that communicate what can be done to improve pagination
     """
-    query_ast = safe_parse_graphql(query_string)
+    query_ast = safe_parse_graphql(query.query_string)
 
-    next_page_ast_with_parameters, remainder_ast_with_parameters = paginate_query_ast(
-        schema_info, query_ast, parameters, page_size
+    next_page_ast_with_parameters, remainder_ast_with_parameters, advisories = paginate_query_ast(
+        schema_info, ASTWithParameters(query_ast, query.parameters), page_size
     )
 
     page_query_with_parameters = QueryStringWithParameters(
         print_ast(next_page_ast_with_parameters.query_ast),
         next_page_ast_with_parameters.parameters,
     )
-    remainder_query_with_parameters = None
-    if remainder_ast_with_parameters is not None:
-        remainder_query_with_parameters = QueryStringWithParameters(
-            print_ast(remainder_ast_with_parameters.query_ast),
-            remainder_ast_with_parameters.parameters,
-        )
+    remainder_queries_with_parameters = tuple(
+        QueryStringWithParameters(print_ast(query.query_ast), query.parameters)
+        for query in remainder_ast_with_parameters
+    )
 
-    return page_query_with_parameters, remainder_query_with_parameters
+    return page_query_with_parameters, remainder_queries_with_parameters, advisories
```

Comment:
Thoughts on making some pleasant dataclass for these?

Reply:
What's wrong with `Tuple`? I don't see a need for an extra layer of nesting here.

Reply:
I guess grouping the first page with the remainder queries into a dataclass will create a nice space to document their relationship once, and cut on documentation needs throughout the codebase each time we use them. But each time we invoke this concept, it's a bit different (sometimes it's query strings, sometime ASTs, etc), that I think we won't get much from it. Not sure if that's what you're looking for.

Reply:
You might be able to make it a generic container that allows it to work the same for both ASTs and for query strings by taking in the type it should use to represent `(query, args)` combinations. And yes, the motivation would be to have a single place that is well-documented and intuitive, rather than a bunch of non-named destructuring at each call site + verbose docstrings everywhere.

Reply:
Added `PageAndRemainder` dataclass. Not fully propagated to the parameterizer. That doesn't change the API, and I'd rather do it in parallel with https://github.com/kensho-technologies/graphql-compiler/pull/741